torch/distributed/_state_dict_utils.py

```diff
@@ -305,28 +370,29 @@ def _create_cpu_state_dict(
     tensors can be placed on pin_memory or share_memory based on the provided arguments.
     """
 
-    if pin_memory and share_memory:
-        raise ValueError(
-            "Cannot allocate both memory on both pin_memory and share_memory"
-        )
-
     def tensor_func(
         obj: torch.Tensor,
         pg: Optional[dist.ProcessGroup],
         device: Optional[torch.device],
-        companion_obj: Any,
+        _: Any,
     ) -> torch.Tensor:
         if len(obj.size()) == 0:
             return torch.tensor(0, dtype=obj.dtype)
 
         if share_memory:
-            return torch.empty(
-                *tuple(companion_obj.size()), dtype=companion_obj.dtype
-            ).share_memory_()
+            t = torch.empty(*tuple(obj.size()), dtype=obj.dtype).share_memory_()
+            if pin_memory:
+                succ = torch.cuda.cudart().cudaHostRegister(
+                    t.data_ptr(),
+                    t.numel() * t.element_size(),
+                    1,  # lines up with 'cudaHostRegisterPortable'
+                )
+                assert succ == 0, f"Pinning shared memory failed with {succ}"
+            return t
+        elif pin_memory:
+            return torch.empty(*tuple(obj.size()), dtype=obj.dtype).pin_memory()
```

Comment:
any particular reason you use the .pin_memory() function here rather than also using the cudart pathway here?  what are the tradeoffs?  IIUC only the cudart pathway works if its also shared, but i'm not sure if .pin_memory is in some way better when its possible to use it.

Reply:
From my understanding (had a hard time tracing down sources), `pin_memory` is significantly faster then calling `cudaHostRegister`. From Sam Gross's explanation my understanding is that `pin_memory` provides a pinned memory cacheing mechanism which avoids directly allocating extra memory, but is incompatible with shared memory.

The only reason to use `cudart` directly is because we don't currently support shared + pinned in the higher level API.

Reply:
Added a docstring warning users to only use this case if the tensor is long lived.